src/Core/src/Credentials/ContainerProvider.php

```diff
@@ -0,0 +1,108 @@
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core\Credentials;
+
+use AsyncAws\Core\Configuration;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Exception\JsonException;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Contracts\HttpClient\Exception\DecodingExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Provides Credentials from the running ECS.
+ *
+ * @see https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/auth/ContainerCredentialsProvider.html
+ */
+final class ContainerProvider implements CredentialProvider
+{
+    private const ENDPOINT = 'http://169.254.170.2';
+    private const ENV_URI = 'AWS_CONTAINER_CREDENTIALS_RELATIVE_URI';
+    private const ENV_TIMEOUT = 'AWS_METADATA_SERVICE_TIMEOUT';
+
+    private $logger;
+
+    private $httpClient;
+
+    private $timeout;
+
+    private $envUri;
+
+    public function __construct(?HttpClientInterface $httpClient = null, ?LoggerInterface $logger = null, float $timeout = 1.0)
+    {
+        $this->logger = $logger ?? new NullLogger();
+        $this->httpClient = $httpClient ?? HttpClient::create();
+        $this->timeout = getenv(self::ENV_TIMEOUT) ?: $timeout;
+        $this->envUri = getenv(self::ENV_URI) ?: '';
+    }
+
+    public function getCredentials(Configuration $configuration): ?Credentials
+    {
+        // fetch credentials from ecs endpoint
+        try {
+            $response = $this->httpClient->request('GET', self::ENDPOINT . $this->envUri, ['timeout' => $this->timeout]);
+            $result = $this->toArray($response);
+
+            if (200 != $response->getStatusCode()) {
+                $this->logger->info('Unexpected instance profile.', ['response_code' => $result['Code']]);
+
+                return null;
+            }
+        } catch (DecodingExceptionInterface $e) {
+            $this->logger->info('Failed to decode Credentials.', ['exception' => $e]);
+
+            return null;
+        } catch (TransportExceptionInterface $e) {
+            $this->logger->info('Failed to fetch Profile from Instance Metadata.', ['exception' => $e]);
+
+            return null;
+        } catch (HttpExceptionInterface $e) {
+            $this->logger->info('Failed to fetch Profile from Instance Metadata.', ['exception' => $e]);
+
+            return null;
+        }
+
+        return new Credentials(
+            $result['AccessKeyId'],
+            $result['SecretAccessKey'],
+            $result['Token'],
+            new \DateTimeImmutable($result['Expiration'])
+        );
+    }
+
+    /**
+     * Copy of Symfony\Component\HttpClient\Response::toArray without assertion on Content-Type header.
```

Comment:
Why? Doesn't the response include the proper Content-Type?

Reply:
I think this is a copy paste from `InstanceProvider`. 

If AWS return an invalid content type, I sugest to move this in a trait. Otherwise, you can delete it

Reply:
So, this is what I get when fetching a request.

> GET /v2/credentials/79ec8795-217b-4bc6-9fee-ba50fb71d87e HTTP/1.1
--
100 | > Host: 169.254.170.2
101 | > User-Agent: curl/7.67.0
102 | > Accept: */*
103 | >
104 | * Mark bundle as not supporting multiuse
105 | < HTTP/1.1 200 OK
106 | < Content-Type: application/json
107 | < X-Rate-Limit-Duration: 1
108 | < X-Rate-Limit-Limit: 40
109 | < X-Rate-Limit-Request-Forwarded-For:
110 | < X-Rate-Limit-Request-Remote-Addr: 169.254.172.2:41774
111 | < Date: Tue, 05 May 2020 18:44:56 GMT
112 | < Content-Length: 1499
113 | <

It's a valid json so I will get rid of it.

Reply:
thanks for the test!! `Content-Type: application/json`is what you need.
You can delete this `toArray` method and use `$response->toArray()` instead